# 4.5 Data Consistency Checks

Table of Contents
1. Importing libraries and data
2. What are Data Consistency Checks and Mixed-Type Data
3. Missing Values
4. Finding Duplicates
5. Task 4.5 Data Cosistency Checks

# 1. Importing Data

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os 

In [2]:
# Importing a Data Set in Jupyter -- df_ords and df_prods

In [3]:
# Importing products data from Original Data folder as df_prods

In [4]:
path = r'C:\Users\peter\Documents\Instacart Basket Analysis 2022'

In [5]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [6]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [7]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [8]:
# Import orders data from Prepared Data folder as df_ords

In [9]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [10]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_days_of_week  \
0           0   2539329        1             1                    2   
1           1   2398795        1             2                    3   
2           2    473747        1             3                    3   
3           3   2254736        1             4                    4   
4           4    431534        1             5                    4   

   order_time  days_since_prior_order  
0           8                     NaN  
1           7                    15.0  
2          12                    21.0  
3           7                    29.0  
4          15                    28.0

In [11]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_days_of_week    order_time  days_since_prior_order  
count         3.421083e+06  3.421083e+06            3.214874e+06  
mean          2.776219e+00  1.345202e+01            1.111484e+01  
std           2.046829e+00  4.226088e+00            9.206737e+00  
min           0.000000e+00  0.000000e+00            0.000000e+00  
25%           1.000000e+00  1.000000e+01            4.000000e+00  
50%           3.000000e+00  1.300000e+01            7.000000e+00  
75%           5.000000e+00  1.600000e+01            1.500000e+01  
max           6.000000e+00  2.300000e+01            3.000000e+01

# 2. What are Data Consistency Checks and Mixed-Type Data

In [12]:
# Finding and addressing mixed data types
# Finding and addressing missing values
# Finding and addressing duplicate records

#  Data prep and data analysis should be two discrete stages in any data project!

In [13]:
# create a dataframe

df_test = pd.DataFrame()

In [14]:
# Create a mixed type column

df_test['mix'] = ['a', 'b', 1, True]

In [15]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [16]:
# Checking for mixed-type columns

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [17]:
# Fixing mixed-type columns

In [18]:
df_test['mix'] = df_test ['mix'].astype('str')

# 3. Missing Values

In [19]:
# Finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [20]:
# Creating a subset to see the 16 missing values under product_name

In [21]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [22]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [23]:
# Missing values cannot be replaced with imputation becasue they are string values 

In [24]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [25]:
df_prods.shape

(49693, 5)

In [26]:
df_prods_clean.shape

(49677, 5)

# 04. Finding Duplicates

In [27]:
# For your Instacart project, you’ll want to look for full duplicates—multiple rows that have the exact same values in every column. This is because single duplicates aren’t actually inconsistencies in your data. Take your "aisle\_id" and "department\_id" columns, for example. The same values occur multiple times throughout these columns, but this doesn’t mean they’re true duplicates 

In [28]:
# looking for duplicates within the dataframe 
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [29]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [30]:
# Duplicates have been identified and can now be deleted

In [31]:
# Checking number of rows in df_prods_clean

df_prods_clean.shape

(49677, 5)

In [32]:
# Drop duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [33]:
df_prods_clean_no_dups.shape

(49672, 5)

In [34]:
# The five duplicates identified using df_dups have been removed. 

# TASK 4.5 Data Cosistency Checks

In [35]:
# 2 --- Run the df.describe() function on your df_prods dataframe. 
# Using your new knowledge about how to interpret the output of this function, 
# share in a markdown cell whether anything about the data looks off or should be investigated further.

In [36]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

Answer: The price max value is 99999, which seems like it could be a mistake and should be checked for accuracy. 

In [37]:
# 3 --- Check for mixed-type data in your df_ords dataframe.

In [38]:
#import orders_wrangled data frame

df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)


In [39]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_days_of_week  \
0           0   2539329        1             1                    2   
1           1   2398795        1             2                    3   
2           2    473747        1             3                    3   
3           3   2254736        1             4                    4   
4           4    431534        1             5                    4   

   order_time  days_since_prior_order  
0           8                     NaN  
1           7                    15.0  
2          12                    21.0  
3           7                    29.0  
4          15                    28.0

In [40]:
# Checking for mixed types in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [41]:
# 4---- If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

Answer: no mixed-type data was found within the df_ords data frame

In [42]:
# 5 ---- Run a check for missing values in your df_ords dataframe.
# In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [43]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [44]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_days_of_week            0
order_time                     0
days_since_prior_order    206209
dtype: int64

In [45]:
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [46]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_days_of_week  \
0                 0   2539329        1             1                    2   
11               11   2168274        2             1                    2   
26               26   1374495        3             1                    1   
39               39   3343014        4             1                    6   
45               45   2717275        5             1                    3   
...             ...       ...      ...           ...                  ...   
3420930     3420930    969311   206205             1                    4   
3420934     3420934   3189322   206206             1                    3   
3421002     3421002   2166133   206207             1                    6   
3421019     3421019   2227043   206208             1                    1   
3421069     3421069   3154581   206209             1                    3   

         order_time  days_since_prior_order  
0                 8                     NaN  
11               11                     NaN  
26               14                     NaN  
39               11                     NaN  
45               12                     NaN  
...             ...                     ...  
3420930          12                     NaN  
3420934          18                     NaN  
3421002          19                     NaN  
3421019          15                     NaN  
3421069          11                     NaN  

[206209 rows x 7 columns]

In [47]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_days_of_week    order_time  days_since_prior_order  
count         3.421083e+06  3.421083e+06            3.214874e+06  
mean          2.776219e+00  1.345202e+01            1.111484e+01  
std           2.046829e+00  4.226088e+00            9.206737e+00  
min           0.000000e+00  0.000000e+00            0.000000e+00  
25%           1.000000e+00  1.000000e+01            4.000000e+00  
50%           3.000000e+00  1.300000e+01            7.000000e+00  
75%           5.000000e+00  1.600000e+01            1.500000e+01  
max           6.000000e+00  2.300000e+01            3.000000e+01

In [48]:
# In a markdown cell, report your findings and propose an explanation for any missing values you find.

Answer: The days_since_prior_order has 206,209 missing values. The missing values here may be due to customers having placed only one order and therefore there is no "days" since their last ordered good. I.e. someone who maybe only ordered a product one time and has not ordered since. 

In [49]:
# 6 ---- Address the missing values using an appropriate method.
# In a markdown cell, explain why you used your method of choice.

The missing observations provide useful data as it would indicate that these customers are ordering through instacart for the very time. Therefore, we should create a new variable that makes note of these missing values. The creation of a new column would be called first_order and this would then indicate why there are missing values within the days_since_prior_order column

In [50]:
# Defining a new data frame for clean data
df_ords_clean = df_ords

In [51]:
df_ords_clean['first_order'] = df_ords['days_since_prior_order'].isnull() == True

In [52]:
df_ords_clean.head()

Unnamed: 0  order_id  user_id  order_number  orders_days_of_week  \
0           0   2539329        1             1                    2   
1           1   2398795        1             2                    3   
2           2    473747        1             3                    3   
3           3   2254736        1             4                    4   
4           4    431534        1             5                    4   

   order_time  days_since_prior_order  first_order  
0           8                     NaN         True  
1           7                    15.0        False  
2          12                    21.0        False  
3           7                    29.0        False  
4          15                    28.0        False

In [53]:
# 7--- Run a check for duplicate values in your df_ords data.
# In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [54]:
df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [55]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_days_of_week, order_time, days_since_prior_order, first_order]
Index: []

In [56]:
# 8 --- Address the duplicates using an appropriate method.
# In a markdown cell, explain why you used your method of choice.

In [57]:
df_ords_dups.shape

(0, 8)

Answer: There are no duplicates. 

In [58]:
# 9 --- Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” 
# folder and give them appropriate, succinct names.

In [59]:
df_ords_clean.to_csv(os.path.join (path, '02 Data', 'Prepared Data', 'orders_cleaned.csv'))

In [60]:
df_prods_clean.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_cleaned.csv'))